test/jdk/java/net/httpclient/http2/PostPutTest.java

```diff
@@ -0,0 +1,116 @@
+
+    // As per jdk.internal.net.http.WindowController.DEFAULT_INITIAL_WINDOW_SIZE
+    final int DEFAULT_INITIAL_WINDOW_SIZE = (64 * 1024) - 1;
+    // Add on a small amount of arbitrary bytes to see if client hangs when receiving RST_STREAM
+    byte[] data = new byte[DEFAULT_INITIAL_WINDOW_SIZE + 10];
+
+    @BeforeTest
+    public void setup() throws Exception {
+        http2TestServer = new Http2TestServer(false, 0);
+        http2TestServer.addHandler(new TestHandler(), "/");
+        http2TestServer.setExchangeSupplier(PostPutExchange::new);
+        http2TestServer.start();
+        uri = new URI("http://" + http2TestServer.serverAuthority() + "/");
+        testLog.println("PostPutTest.setup(): Test Server URI: " + uri);
+    }
+
+    @AfterTest
+    public void teardown() {
+        testLog.println("PostPutTest.teardown(): Stopping server");
+        http2TestServer.stop();
+        data = null;
+    }
+
+    @DataProvider(name = "variants")
+    public Object[][] variants() {
+        HttpRequest over64kPost = HttpRequest.newBuilder().version(HTTP_2).POST(ofByteArray(data)).uri(uri).build();
+        HttpRequest over64kPut = HttpRequest.newBuilder().version(HTTP_2).PUT(ofByteArray(data)).uri(uri).build();
+
+        return new Object[][] {
+                { over64kPost, "POST data over 64k bytes" },
+                { over64kPut, "PUT data over 64k bytes" }
+        };
+    }
+
+    public HttpRequest getWarmupReq() {
+        return HttpRequest.newBuilder()
+                .GET()
+                .uri(uri)
+                .build();
+    }
+
+    @Test(dataProvider = "variants")
+    public void testOver64kPUT(HttpRequest req, String testMessage) {
+        testLog.println("PostPutTest: Performing test: " + testMessage);
+        HttpClient hc = HttpClient.newBuilder().version(HTTP_2).build();
+        hc.sendAsync(getWarmupReq(), HttpResponse.BodyHandlers.ofString()).join();
+        hc.sendAsync(req, HttpResponse.BodyHandlers.ofString()).join();
+        /*
+            If this test fails in timeout, it is likely due to one of two reasons:
+              - The responseSubscriber is null, so no incoming frames are being processed by the client
+                (See Stream::schedule)
+              - The test server is for some reason not sending a RST_STREAM with the NO_ERROR flag set after
+                sending an empty DATA frame with the END_STREAM flag set.
+        */
+    }
+
+    private static class TestHandler implements Http2Handler {
+
+        @Override
+        public void handle(Http2TestExchange exchange) throws IOException {
+            // The input stream is not read in this bug as this will trigger window updates for the server concerns
+            // the case where no updates are sent and the server instead tells the client to abort the transmission.
+            exchange.sendResponseHeaders(200, 0);
+        }
+    }
+
+    static class PostPutExchange extends Http2TestExchangeImpl {
```

Comment:
Can we make the HTTP2 server always reset the stream if the configured Http2Handler doesn't consume the entire input?
There's even a TODO for that:
https://github.com/openjdk/jdk/blob/master/test/jdk/java/net/httpclient/lib/jdk/httpclient/test/lib/http2/BodyInputStream.java#L133

Reply:
Update on this, the issue here is making sure that the reset frame on the server side is triggered in the correct manner for an individual BodyInputStream instance. This can be done by checking various conditions such as the stream having unprocessed/read frames in it's Queue (in other words checking if the entire input has not been consumed highlighted by @djelinski above) and making sure that a reset frame is not incorrectly sen after receiving a connection preface, settings etc.

I've almost got this sorted though, so hopefully new changes here shortly